force compiling std from source if modified

onur-ozkan opened the pull request:
This allows the standard library to be compiled even with `download-rustc` enabled. Which means it's no longer a requirement to compile `rustc` in order to compile `std`.

Ref. https://github.com/rust-lang/rust/pull/127322#discussion_r1666418280.
Blocker for https://github.com/rust-lang/rust/pull/122709.

onur-ozkan commented:
r? bootstrap

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10011130778/job/27673890907) [(plain)](https://github.com/rust-lang/rust/commit/b4515eeb8189dab19f082176480987eca0e19f99/checks/27673890907/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:3aacb9c90579defe09351ac5e8ee504359f8054da6326ff19038f1b7c90e3cb2aafe33685c6d9b76ee8d2ccbd187ca80c46ab5380485abdd8c0ce7d69cd8d8fd:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.94s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---

---- core::builder::tests::alias_and_path_for_library stdout ----
thread 'core::builder::tests::alias_and_path_for_library' panicked at src/core/builder/tests.rs:204:5:
assertion `left == right` failed
  left: [Std { target: A-A, compiler: Compiler { stage: 0, host: A-A }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }, Std { target: A-A, compiler: Compiler { stage: 1, host: A-A }, crates: [], force_recompile: true, extra_rust_args: [], is_for_mir_opt_tests: false }]
 right: [Std { target: A-A, compiler: Compiler { stage: 0, host: A-A }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }, Std { target: A-A, compiler: Compiler { stage: 1, host: A-A }, crates: [], force_recompile: false, extra_rust_args: [], is_for_mir_opt_tests: false }]

---- core::builder::tests::defaults::build_cross_compile stdout ----
thread 'core::builder::tests::defaults::build_cross_compile' panicked at src/core/builder/tests.rs:270:9:
assertion failed: `(left == right)`
```

</details>

Kobzol reviewed (commented):
It seems a bit unintuitive to me that `x build library` uses this logic, but `x build` doesn't. Couldn't this use the same logic as `rustdoc`, which checks if the library isn't modified?

onur-ozkan commented:
> It seems a bit unintuitive to me that `x build library` uses this logic, but `x build` doesn't. Couldn't this use the same logic as `rustdoc`, which checks if the library isn't modified?

We already compile `library` if it's changed. This is for compiling `library` only, while using `rustc` from CI. Doing this by default with `x build` would kill the purpose of `download-rustc`.

Kobzol commented:
I tested it locally, seems to work fine. We should however somehow document (?) that this only makes sense for `rust.download-rustc = true`, with `if-unchanged` it still rebuilds the whole sysroot starting from stage 0.

Mark-Simulacrum commented on src/bootstrap/src/core/build_steps/compile.rs:
```diff
@@ -136,7 +142,8 @@ impl Step for Std {
             crates,
-            force_recompile: false,
+            // Force compiling from source as it was explicitly called with `x build library (or std/sysroot)`.
+            force_recompile: !run.builder.paths.is_empty(),
```
It seems like this will be recompilation even if e.g. `x.py build compiler` is passed? I'm not sure that gating on just any path is the right call. Shouldn't this use something like "is std modified?" perhaps?

onur-ozkan replied:
> It seems like this will be recompilation even if e.g. x.py build compiler is passed? I'm not sure that gating on just any path is the right call.

No, that shouldn't happen.

> Shouldn't this use something like "is std modified?" perhaps?

We can do that but it's not necessary in my opinion. If user explicitly asks for library build, we can do it right away without checking if "is std modified?".

Mark-Simulacrum replied:
I think it's pretty confusing -- certainly against my expectations -- that the same source tree can produce *different* results depending on the command (x.py build vs. x.py build library, for example). I'd expect that to always yield the same results, though the path-less (`x.py build`) would perhaps build more.

If we choose whether to rebuild std based on the paths provided, that violates this basic assumption that I think we have so far upheld.

onur-ozkan replied:
[Changed](https://github.com/rust-lang/rust/pull/127974/commits/bf7fb649a24320cd78758944f0274f8483c7478d) into `if modified` logic.

onur-ozkan committed:
set `force_recompile: true` if library is modified

This allows the standard library to be compiled even with `download-rustc` enabled.
Which means it's no longer a requirement to compile `rustc` in order to compile `std`.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan committed:
update download-rustc documentation

Signed-off-by: onur-ozkan <work@onurozkan.dev>

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 6fcc630e56263d7d17a37a77202717f1c83c5b2a has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 6fcc630e56263d7d17a37a77202717f1c83c5b2a -->
<!-- homu: {"type":"Approved","sha":"6fcc630e56263d7d17a37a77202717f1c83c5b2a","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
